wandb/sdk/interface/interface.py

```diff
@@ -447,15 +447,79 @@ def publish_link_artifact(
     def _publish_link_artifact(self, link_artifact: pb.LinkArtifactRecord) -> None:
         raise NotImplementedError
 
+    def _make_proto_use_artifact(
+        self,
+        use_artifact: pb.UseArtifactRecord,
+        job_name: str,
+        job_info: Dict[str, Any],
+        metadata: Dict[str, Any],
+    ) -> pb.UseArtifactRecord:
+        use_artifact.partial.job_name = job_name
+        use_artifact.partial.source_info._version = job_info.get("_version", "")
+        use_artifact.partial.source_info.source_type = job_info.get("source_type", "")
+        use_artifact.partial.source_info.runtime = job_info.get("runtime", "")
+
+        # construct JobSourceDict.source
+        if job_info.get("source_type") == "artifact":
+            use_artifact.partial.source_info.source.artifact.artifact = job_info.get(
+                "source", {}
+            ).get("artifact", "")
+            use_artifact.partial.source_info.source.artifact.entrypoint.extend(
+                job_info.get("source", {}).get("entrypoint", [])
+            )
+            use_artifact.partial.source_info.source.artifact.notebook = job_info.get(
+                "source", {}
+            ).get("notebook", False)
+        elif job_info.get("source_type") == "repo":
+            use_artifact.partial.source_info.source.git.git_info.remote = metadata.get(
+                "git", {}
+            ).get("remote", "")
+            use_artifact.partial.source_info.source.git.git_info.commit = metadata.get(
+                "git", {}
+            ).get("commit", "")
+            use_artifact.partial.source_info.source.git.entrypoint.extend(
+                metadata.get("entrypoint", [])
+            )
+            use_artifact.partial.source_info.source.git.notebook = metadata.get(
+                "notebook", False
+            )
+        elif job_info.get("source_type") == "image":
+            use_artifact.partial.source_info.source.image.image = metadata.get(
+                "image", ""
+            )
+
+        return use_artifact
+
     def publish_use_artifact(
         self,
         artifact: "Artifact",
     ) -> None:
         assert artifact.id is not None, "Artifact must have an id"
+
         use_artifact = pb.UseArtifactRecord(
-            id=artifact.id, type=artifact.type, name=artifact.name
+            id=artifact.id,
+            type=artifact.type,
+            name=artifact.name,
         )
 
+        if "_partial" in artifact.metadata:
+            # Download source info from logged partial job artifact
+            job_info = {}
+            try:
+                path = artifact.get_path("wandb-job.json").download()
```

Comment:
Let's put a TODO here to indicate that we'd prefer to do this all in the internal process.